godep/client.go

```diff
@@ -68,24 +78,62 @@ type Client struct {
 	store ClientStorage
 
 	// an HTTP client that handles DEP API authentication and session management
-	client *http.Client
+	client  depclient.Doer
+	errHook func(ctx context.Context, err error) error
+}
+
+// ClientOption defines the functional options type for NewClient.
+type ClientOption func(*Client)
+
+// WithMiddleware applies a middleware function to the client's HTTP client.
+// It can be useful to transform the HTTP request just before it is sent, or
+// to intercept the HTTP response and error before they are returned to the
+// caller.
+func WithMiddleware(mw func(depclient.Doer) depclient.Doer) ClientOption {
+	return func(c *Client) {
+		c.client = mw(c.client)
+	}
+}
+
+// WithErrHook installs a hook function that is called with the error resulting
+// from any request, after transformation of the response's body to an
+// HTTPError if needed. It gets called even in case of success, in that case
+// with a nil error. It can return a new error to be returned by the Client, or
+// the original error.
+func WithErrHook(hook func(ctx context.Context, err error) error) ClientOption {
```

Comment:
Why should it be called even when there is a success? Typically error callbacks are only called if there is a non nil error.

Reply:
`WithAfterHook` or `WithResultHook` maybe? `After` is probably straightforward enough, similar to how `urfave/cli` supports "after" command hooks.

Reply:
+1 for `WithAfterHook`